### PR TITLE
[firebase_auth] Clean up listener when detached and channel is null

### PR DESCRIPTION
**Pull request by gching**
## Description

* Remove / clean up listener when `onDetachedFromEngine` is called

### Background
* This happens when `onDetachedFromEngine` is called and when we have a at least one listener in `authStateListeners`, cause we do not remove / clean up listeners when detached, and therefore, when the listener is called after being detached, it will be `null`
* Just to highlight, this only started happening since `0.16.0` cause the package was upgraded to Android v2 embedding.

### Implementation details
* I created `PluginAuthStateListener` that implements `FirebaseAuth.AuthStateListener` that includes a `remove` method that can be called when `onDetachedFromEngine` occurs. Previously, I opted to remove the listener in the listener itself, but I couldn't really test this
* There was no way for me to mock `FirebaseApp` so for tests I had to include `Roboletric`
* I've changed the access for `authStateListeners` to be package-private so it can be accessed for tests

## Related Issues
* Fixes (hopefully) #2444. It worked for me :> but lets get others to test. However, `channel` is only set to `null` when `onDetachedFromEngine` is called so this should be the only reason why it is `null`
* **IMPORTANT** Looking into it more, it seems like some plugins (`cloud_firestore`) has possibly the same issue (ex. #2517, #2526). I actually worked on this before seeing the comments in #2526, so let me know if this is the wrong way of implementing this or if not, I can make separate PRs to address the issue for different plugins. (cc @cyanglaz @lidongze91)

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 1]
```diff
@@ -700,6 +700,12 @@ private void handleStartListeningAuthState(
         new FirebaseAuth.AuthStateListener() {
           @Override
           public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
+            if (channel == null) {
```
[the remark this answers is not in the document]
Comment (gching) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:703]: Just to confirm, you like tests for this on the Java side? Don't think I can write tests for this at the Dart level and I tried to look for tests for `FirebaseAuthPlugin`. Is there a place for these tests?

[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:703]: yeah, java unit test should be fine

[the remark this answers is not in the document]
Comment (gching) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:703]: Where are the current tests? Just trying to look for it!

[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:703]: I don't think any junit tests are set up in this plugin. You can take https://github.com/flutter/plugins/blob/master/packages/image_picker/image_picker/example/android/app/src/test/java/io/flutter/plugins/imagepicker/ImagePickerPluginTest.java as an example

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 2]
```diff
@@ -62,7 +63,7 @@
 public class FirebaseAuthPlugin implements MethodCallHandler, FlutterPlugin, ActivityAware {
   // Only set registrar for v1 embedder.
   private PluginRegistry.Registrar registrar;
-  private SparseArray<AuthStateListener> authStateListeners;
+  SparseArray<PluginAuthStateListener> authStateListeners;
```
Comment (gching) [inline packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:66]: Had to expose this for tests, is this fine?

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 3]
```diff
@@ -889,4 +905,39 @@ private void reportException(Result result, @Nullable Exception exception) {
       result.error("ERROR_UNKNOWN", "An unknown error occurred.", null);
     }
   }
+
+  class PluginAuthStateListener implements AuthStateListener {
```
Comment (gching) [inline packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:909]: Created this class to help with removing it on `onDetachedFromEngine`

File: packages/firebase_auth/firebase_auth/example/android/app/src/test/java/io/flutter/plugins/firebaseauth/FirebaseAuthPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  @Mock FlutterPlugin.FlutterPluginBinding mockFlutterPluginBinding;
+  @Mock MethodChannel.Result mockResult;
+
+  private String testFirebaseAppName = "[DEFAULT]";
+
+  FirebaseAuthPlugin plugin;
+
+  @BeforeClass()
+  public static void BeforeClass() {
+    FlutterMain.setIsRunningInRobolectricTest(true);
```
Comment (gching) [inline packages/firebase_auth/firebase_auth/example/android/app/src/test/java/io/flutter/plugins/firebaseauth/FirebaseAuthPluginTest.java:40]: I've seen you use this in the `google_maps` plugin but not in `image_picker`. I cannot figure out how to run tests without calling this and I know it is deprecated :(

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java (hunk 2 above)
Comment (gching) [inline packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:66]: Had to expose this for tests

**Comment by gching** [comment]
@cyanglaz  Also, I opted to not change and not call `listener.remove` in `handleStopListeningAuthState` https://github.com/FirebaseExtended/flutterfire/blob/07726177f3ee9f68c6f0d430a39e308551f3f9db/packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java#L741 cause it felt weird to not use the passed in `FirebaseAuth` instance. Let me know what you think about this.

**Comment by gching** [comment]
@bparrishMines can you review as well?

**Comment by gching** [comment]
@cyanglaz @bparrishMines Just bumping this so I can get feedback on this

**Comment by cyanglaz** [comment]
Will do it today. Might be a little late tho. Hope that’s ok

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java (hunk 2 above)
Comment (cyanglaz) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:66]: What do you think about keeping this private and adding a @VisibleForTestings method: `GetAuthStateListeners` to use for testing?

File: packages/firebase_auth/firebase_auth/example/android/app/src/test/java/io/flutter/plugins/firebaseauth/FirebaseAuthPluginTest.java (hunk 4 above)
Comment (cyanglaz) [reply packages/firebase_auth/firebase_auth/example/android/app/src/test/java/io/flutter/plugins/firebaseauth/FirebaseAuthPluginTest.java:40]: I don't recall entirely but I think this is required for certain robotic methods. If the test fails without this, then you probably need it.

File: packages/firebase_auth/firebase_auth/example/android/gradle.properties [hunk 5]
```diff
@@ -2,3 +2,4 @@ org.gradle.jvmargs=-Xmx1536M
 android.enableR8=true
 android.useAndroidX=true
 android.enableJetifier=true
+android.enableUnitTestBinaryResources=true
```
Comment (cyanglaz) [inline packages/firebase_auth/firebase_auth/example/android/gradle.properties:5]: nits: line EOF

File: packages/firebase_auth/firebase_auth/example/android/app/src/test/java/io/flutter/plugins/firebaseauth/FirebaseAuthPluginTest.java (hunk 4 above)
Comment (gching) [reply packages/firebase_auth/firebase_auth/example/android/app/src/test/java/io/flutter/plugins/firebaseauth/FirebaseAuthPluginTest.java:40]: @cyanglaz Gotcha!

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java (hunk 2 above)
Comment (gching) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:66]: I'll try this :)

**Comment by Ehesp** [comment]
This is now handled in the new upgrade. Thanks!

**Pull request closed without merging**